core, core/vm, cmd/disasm: unify procedures for disassembling evm code

wuestholz opened the pull request:
This cleans up and unifies several spread out procedures for disassembling and pretty-printing evm code as suggested by @fjl in response to the following earlier PR: https://github.com/ethereum/go-ethereum/pull/3491.

fjl commented on core/asm/asm.go:
```diff
@@ -0,0 +1,78 @@
+	for pc := uint64(0); pc < uint64(len(script)); pc++ {
+		op := vm.OpCode(script[pc])
+		switch op {
+		case vm.PUSH1, vm.PUSH2, vm.PUSH3, vm.PUSH4, vm.PUSH5, vm.PUSH6, vm.PUSH7, vm.PUSH8, vm.PUSH9, vm.PUSH10, vm.PUSH11, vm.PUSH12, vm.PUSH13, vm.PUSH14, vm.PUSH15, vm.PUSH16, vm.PUSH17, vm.PUSH18, vm.PUSH19, vm.PUSH20, vm.PUSH21, vm.PUSH22, vm.PUSH23, vm.PUSH24, vm.PUSH25, vm.PUSH26, vm.PUSH27, vm.PUSH28, vm.PUSH29, vm.PUSH30, vm.PUSH31, vm.PUSH32:
```
You can use `op >= vm.PUSH1 && op <= vm.PUSH32` here if you change this
to a bool switch (with no switch value).

fjl commented on core/asm/asm.go:
```diff
@@ -0,0 +1,78 @@
+)
+
+// Apply function to each disassembled EVM instruction.
+func ForEachDisassembledInstruction(script []byte, fun func(uint64, vm.OpCode, []byte)) error {
```
IMHO it would be better to have a real iterator here. The API would be as follows:

```go
it := asm.NewIterator(code)
for it.Next() { // Next returns true until there are no instructions left.
    // The PC, Op, Arg fields contain values for the current instruction.
    fmt.Println(it.PC, it.Op, it.Arg)
}
fmt.Println(it.Error()) // Error returns any error that occurred in Next.
```

fjl commented on core/asm/asm.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Pretty-print all disassembled EVM instructions to stdout.
+func PrettyPrintDisassembledInstructions(code string) error {
```
The names of these functions could be shorter. Maybe "Disassemble" and "WriteDisassembly"?

fjl commented on core/asm/asm.go:
```diff
@@ -0,0 +1,78 @@
+// Copyright 2014, 2017 The go-ethereum Authors
```
Only one year is needed. We usually put the year in which the file was first introduced.

fjl commented:
The idea would be that Next returns false when an invalid push is decoded.

```go
func (it *Iterator) Next() bool {
    if it.err != nil || it.offset == len(it.code) {
        return false      
    }
    op := code[it.offset]
    switch {
    case op >= vm.PUSH1 && op <= vm.PUSH32:
        if invalidpush {
            it.err = fmt.Errorf("invalid push etc.")
            return false
        }
        ...
    }
    return true
}
```

wuestholz commented:
@fjl Thank you for the clarification and the suggestion! That's more or less what I had in mind as well. I came up with the following code:

```go
// Iterator for disassembled EVM instructions
type instructionIterator struct {
	code  []byte
	pc    uint64
	arg   []byte
	op    vm.OpCode
	error error
}

// Create a new instruction iterator.
func NewInstructionIterator(code []byte) *instructionIterator {
	it := new(instructionIterator)
	it.code = code
	it.pc = 0
	it.error = nil
	it.arg = nil
	return it
}

// Returns true if there is a next instruction and moves on.
func (it *instructionIterator) Next() bool {
	if it.error != nil || uint64(len(it.code)) <= it.pc {
		// We previously reached an error or the end.
		return false
	}

	if it.arg == nil {
		// We are calling this procedure for the first time.
		it.pc = 0
	} else {
		// We are moving to the next instruction.
		it.pc += uint64(len(it.arg)) + 1
	}

	if uint64(len(it.code)) <= it.pc {
		// We reached the end.
		return false
	}

	it.op = vm.OpCode(it.code[it.pc])
	if it.op.IsPush() {
		a := uint64(it.op) - uint64(vm.PUSH1) + 1
		u := it.pc + 1 + a
		if uint64(len(it.code)) <= it.pc || uint64(len(it.code)) < u {
			it.error = fmt.Errorf("incomplete push instruction at %v", it.pc)
			return false
		}
		it.arg = it.code[it.pc+1 : u]
	} else {
		it.arg = make([]byte, 0)
	}
	return true
}

// Returns any error that may have been encountered.
func (it *instructionIterator) Error() error {
	return it.error
}

// Returns the PC of the current instruction.
func (it *instructionIterator) PC() uint64 {
	return it.pc
}

// Returns the opcode of the current instruction.
func (it *instructionIterator) Op() vm.OpCode {
	return it.op
}

// Returns the argument of the current instruction.
func (it *instructionIterator) Arg() []byte {
	return it.arg
}
```

I prefer the previous design. However, if you prefer this, I'd be happy to change it. Alternatively, I could add this and implement `ForEachDisassembledInstruction` using the iterator to support both APIs.

wuestholz commented:
@fjl Did you have a chance to look at the iterator code above? Please let me know if committing it makes it easier for you to review it.

fjl commented:
Sorry, didn't have time to look yet

obscuren commented:
@fjl update please

wuestholz commented:
@obscuren Would you like me to rebase the fork?

obscuren commented:
@wuestholz I don't think @fjl has the time to review, I'll pick this up (sorry for the delay). What I think would be best is if we merge the `disasm` tool to the `evm` utility much as I've done with https://github.com/ethereum/go-ethereum/pull/3686

obscuren commented:
@wuestholz The iterator seems about right to be as well and imo is better than your previous code. Would you like to work on the integration part in to the `evm` utility as well?

obscuren commented:
At some point (though not needed for this PR) it would be nice if the disasm tool becomes a 2-stage disassembler where `JUMPDEST`s are parsed and labeled such that

```
PUSH1 03
JUMP
JUMPDEST
```

becomes

```
jump @label_1
label_1:
```

wuestholz commented:
@obscuren Thank you very much for the feedback. I have rebased and added the iterator.

I'd be happy to integrate `disasm` into `evm`. However, I'd prefer to create a separate PR for that once your PR and this one have been merged. What do you think?

obscuren commented on core/asm/asm.go:
```diff
@@ -0,0 +1,138 @@
+func NewInstructionIterator(code []byte) *instructionIterator {
+	it := new(instructionIterator)
+	it.code = code
+	it.pc = 0
```
There's no need to initialise `pc`, `error`, `arg`. These are by default initialised to their default value i.e. `0` and `nil`.

obscuren commented on core/asm/asm.go:
```diff
@@ -0,0 +1,138 @@
+		return false
+	}
+
+	if it.arg == nil {
```
Why do you need to check if `arg` is `nil` here and reset the `pc`?

obscuren commented on core/asm/asm.go:
```diff
@@ -0,0 +1,138 @@
+		}
+		it.arg = it.code[it.pc+1 : u]
+	} else {
+		it.arg = make([]byte, 0)
```
If you can remove the `arg == nil` check on line 53 you can set it to `nil` here instead of allocating a new empty silce.

obscuren commented on core/asm/asm.go:
```diff
@@ -0,0 +1,138 @@
+			instrs = append(instrs, fmt.Sprintf("%06v: %v\n", it.PC(), it.Op()))
+		}
+	}
+	if it.Error() != nil {
```
Please change to

```
if err := it.Error(); err != nil {
    return nil, err
}
```

wuestholz commented:
@obscuren Thank you very much for the feedback. I have addressed everything. The `it.arg == nil` check above was necessary to distinguish the first iteration from the subsequent ones, where we need to increment `pc`. Since it.arg can be nil now at other times as well I have introduced a boolean `started` to express this more explicitly.
